contracts/sui/xcall/sources/main.move

```diff
@@ -0,0 +1,547 @@
+        response_code: u8
+    }
+    /***************/
+   
+    
+    fun init(ctx: &mut TxContext) {
+        let admin = xcall_state::create_admin_cap(ctx);
+        let storage = xcall_state::create_storage(
+           CURRENT_VERSION,
+            &admin,
+            ctx
+        );
+       xcall_state::share(storage);
+       xcall_state::transfer_admin_cap(admin,ctx);
+    }
+
+    public fun register_dapp<T: drop>(self:&Storage,
+        witness: T,
+        ctx: &mut TxContext
+    ):IDCap {
+        self.create_id_cap(ctx)
+    }
+
+    entry public fun configure_nid(self:&mut Storage,admin:&AdminCap, net_id:String,ctx: &mut TxContext){
+        if(!(net_id == b"".to_string())){
+            xcall_state::set_net_id(self,net_id);
+        }
+    }
+
+    entry public fun get_nid(self: &mut Storage): String{
+        xcall_state::get_net_id(self)
+    }
+
+
+    entry public fun register_connection(self:&mut Storage,admin:&AdminCap,net_id:String,connection_id:String,relayer:address,ctx: &mut TxContext){
+        self.enforce_version(CURRENT_VERSION);
+        self.set_connection(net_id,connection_id);
+        let cap= xcall_state::new_conn_cap(self.get_id(),connection_id,ctx);
+        xcall_state::transfer_conn_cap(cap,relayer,ctx);
+        register(self.get_connection_states_mut(),connection_id,ctx);
+    }
+
+    public fun admin(self:&mut Storage):ID{
+        xcall_state::get_admin(self)
+    }
+
+    public fun get_fee_handler(self:&mut Storage):address{
+        xcall_state::get_protocol_fee_handler(self)
+    }
+
+    fun get_connection_fee(self:&Storage,connection_id:String,net_id:String, need_response:bool ):u64{
+        connections::get_fee(
+            xcall_state::get_connection_states(self),
+            connection_id,
+            net_id,
+            need_response
+        )
+    }
+
+    entry public fun get_fee(self:&mut Storage, net_id:String, rollback:bool):u64{
+        let mut fee = xcall_state::get_protocol_fee(self);
+        let connection = xcall_state::get_connection(self,net_id);
+        fee = fee + get_connection_fee(self, connection, net_id, rollback);
+        fee
+    }
+
+    entry public fun get_fee_sources(self:&Storage, net_id:String, rollback:bool, sources:vector<String>):u64{
+        let mut fee = xcall_state::get_protocol_fee(self);
+
+        if(isReply(self,net_id,sources) && !rollback){
+            return 0
+        };
+
+        let mut i = 0;
+        while(i < vector::length(&sources)){
+            let source = vector::borrow(&sources, i);
+            fee = fee + get_connection_fee(self,*source, net_id, rollback);
+            i=i+1
+        };
+
+        fee
+    }
+
+
+    fun send_call_inner(self:&mut Storage,fee:Coin<SUI>,from:NetworkAddress,to:NetworkAddress,envelope:XCallEnvelope,ctx: &mut TxContext):Coin<SUI>{
+        self.enforce_version(CURRENT_VERSION);
+        let sequence_no=get_next_sequence(self);
+        let rollback=envelope::rollback(&envelope);
```

Comment:
This is not used anywhere?

Reply:
send_call_inner function? this is used in send_call, or you mean anything else?

Reply:
Sorry,  rollback=envelope::rollback(&envelope); this extraction of rollback